examples/blockchain/src/network.rs

```diff
@@ -233,130 +95,200 @@ impl aleph_bft::Network<NetworkData> for Network {
-                ttl: Duration::from_secs(6 * 60),
-                query_interval: Duration::from_millis(100),
-            };
-            let rq_rp = RequestResponse::new(
-                GenericCodec {},
-                iter::once((ALEPH_PROTOCOL_NAME.as_bytes().to_vec(), protocol_support)),
-                rr_cfg,
-            );
-
-            let mdns = Mdns::new(mdns_config).await?;
-            let behaviour = Behaviour {
-                rq_rp,
-                mdns,
-                peers: vec![],
-                peer_by_index: HashMap::new(),
-                consensus_tx: msg_for_store,
-                block_tx: block_to_data_io_tx,
-                node_ix,
-            };
-            SwarmBuilder::new(transport, behaviour, local_peer_id)
-                .executor(Box::new(|fut| {
-                    tokio::spawn(fut);
-                }))
-                .build()
-        };
-
-        swarm.listen_on("/ip4/0.0.0.0/tcp/0".parse()?)?;
 
         let network = Network {
             msg_to_manager_tx,
             msg_from_manager_rx: msg_from_store,
         };
 
         let network_manager = NetworkManager {
-            swarm,
+            id,
+            address,
+            addresses,
+            bootnodes,
+            listener,
+            consensus_tx: msg_for_store,
             consensus_rx: msg_to_manager_rx,
+            block_tx: block_to_data_io_tx,
             block_rx: block_from_data_io_rx,
         };
 
         Ok((
-            network,
             network_manager,
+            network,
             block_from_data_io_tx,
             block_to_data_io_rx,
             msg_for_network,
             msg_from_manager,
         ))
     }
-}
 
-impl NetworkManager {
+    fn recipient_to_addresses(&self, recipient: Recipient) -> HashMap<usize, Address> {
+        let mut addr: HashMap<usize, Address> = HashMap::new();
+        match recipient {
+            Recipient::Node(n) => {
+                let n: usize = n.into();
+                assert!(n < self.addresses.len());
+                if n != self.id {
+                    if let Some(a) = &self.addresses[n] {
+                        addr.insert(n, a.clone());
+                    }
+                }
+            }
+            Recipient::Everyone => {
+                for n in 0..self.addresses.len() {
+                    if n != self.id {
+                        if let Some(a) = &self.addresses[n] {
+                            addr.insert(n, a.clone());
+                        }
+                    };
+                }
+            }
+        }
+        addr
+    }
+
+    fn reset_dns(&mut self, n: usize) {
+        if !self.bootnodes.contains_key(&(n as u32)) {
+            error!("Reseting address of node {}", n);
+            self.addresses[n] = None;
+        }
+    }
+
+    fn send(&mut self, message: Message, recipient: Recipient) {
+        let addr = self.recipient_to_addresses(recipient);
```

Comment:
I'm not sure this function does much. If you had addresses in a `HashMap` you could just match here and have one simple get instead of what this function does in case of `Recipient::Node`. In case of `Recipient::Everyone` you could just iterate over items

Reply:
yeah, but still it's ~15 lines of code, so I'd leave it as a separate function

Reply:
yes :D